Support global.json's rollForward latest* variants

js6pak opened the pull request:
**Description:**
This PR is a cleaned up version of #481, which fixes 2 issues:
1. In the original `npm run build` wasn't rerun so `dist` is stale there
2. The tests that were added never passed successfully

There are a few additional commits in this PR:
1. https://github.com/actions/setup-dotnet/commit/babb40fcf451e6f5cae684f188c5dbbeb41a58a8 which fixes the 6 minute delay on windows runners caused by `clear-toolcache.ps1`
2. https://github.com/actions/setup-dotnet/commit/d9dae35030614ced7b174b151463481173bfe2a3 makes sure tests run even before the PR workflow is manually accepted, this can be dropped if needed

**Related issue:**
Fixes #448, closes #481

**Check list:**
- [ ] Mark if documentation changes are required.
- [x] Mark if tests were added or updated to cover the changes.

js6pak committed:
Remove branch constraint on workflows used for testing

This means you are going to be notified of CI issues quicker, especially if you have to wait for a confirmation before running the workflows on your PR.

js6pak committed:
Rename toolcache directory instead of moving to tmp

This further enhances time savings seen in #213, down from a few minutes to seconds.

js6pak commented:
@HarithaVattikuti

js6pak commented:
@HarithaVattikuti bump

v-priyagupta108 commented:
Hi @js6pak,

Thank you for this contribution, this fills an important gap in `global.json` support! Here are a few review points:

1. **Rebase onto latest `main`.** This will ensure your changes are up-to-date with the current codebase.

2. **Revert the workflow trigger changes.** The branch filter removals across the CI workflow files (`.github/workflows/*.yml`) appear unrelated to this feature. Please revert those so they remain as-is.

Once those are addressed, happy to take another look. Thanks!

js6pak committed:
Support global.json's rollForward latest* variants

js6pak committed:
Remove e2e-test's dependency on Microsoft.NET.Test.Sdk

The newer versions are net8.0/net462+ only, making older .NET sdks more annoying to test

js6pak commented:
@priyagupta108 I've rebased the PR, so you can take a look. Workflow trigger changes are needed for external contributors to test whether their changes pass CI, so I'd recommend keeping them, but if you want to drop it, I'd do it after making sure you don't want any other changes to the actual feature (so I can test that CI works after requested changes :v).

js6pak commented:
@priyagupta108 Note the `Basic validation` workflow is broken on the `main` branch: https://github.com/actions/setup-dotnet/actions/workflows/basic-validation.yml?query=branch%3Amain, so it's unrelated to this PR.

v-priyagupta108 commented:
Hi @js6pak, thanks for the quick rebase!

A few follow-up items to address:

1. **`__tests__/e2e-test-csproj` changes:** Migrating the entire test project from MSTest to Newtonsoft.Json just to support a single `latestMinor` test case feels like a significant change, especially since that test depends on .NET Core 3.1, which has been EOL since December 2022. Would you consider reverting the migration and dropping the `latestMinor` test instead?

2. **`actions/checkout` version:** Some of the new jobs added in `e2e-tests.yml` still use `actions/checkout@v3`. Could these be updated to `actions/checkout@v6` to match the rest of the workflow?

3. **Branch filters:** The current branch filters (`main`, `releases/*`) are intentional because they prevent CI from running on every push to every branch, which helps keep workflow costs and noise low for the repo. Could you please drop this change from the PR once your testing is complete?

Once those are addressed, happy to continue the review. Thanks!

js6pak committed:
Revert "Remove e2e-test's dependency on Microsoft.NET.Test.Sdk"

This reverts commit 14839a662649cdea79ccecc3a7ce9a91386185b4.

js6pak committed:
Don't run unsupported versions in global.json tests

js6pak committed:
Revert "Remove branch constraint on workflows used for testing"

This reverts commit d9dae35030614ced7b174b151463481173bfe2a3.

js6pak commented:
> A few follow-up items to address:
> 
> 1. **`__tests__/e2e-test-csproj` changes:** Migrating the entire test project from MSTest to Newtonsoft.Json just to support a single `latestMinor` test case feels like a significant change, especially since that test depends on .NET Core 3.1, which has been EOL since December 2022. Would you consider reverting the migration and dropping the `latestMinor` test instead?
> 2. **`actions/checkout` version:** Some of the new jobs added in `e2e-tests.yml` still use `actions/checkout@v3`. Could these be updated to `actions/checkout@v6` to match the rest of the workflow?
> 3. **Branch filters:** The current branch filters (`main`, `releases/*`) are intentional because they prevent CI from running on every push to every branch, which helps keep workflow costs and noise low for the repo. Could you please drop this change from the PR once your testing is complete?
> 
> Once those are addressed, happy to continue the review. Thanks!

Thanks for catching 2., I disagree with 1. and 3. but addressed them.

nikita-bestuzhev-ct commented:
@HarithaVattikuti, @priyagupta108, hello!

JFYI, the following `global.json`:
```
{
  "sdk": {
    "version": "8.0",
    "rollForward": "latestFeature"
  }
}
```
now fails with:
`Error: Cannot read properties of undefined (reading 'substring')`

See https://github.com/actions/setup-dotnet/pull/538/changes#diff-83766f6dc1065d750be380c25c707c33a1516e036ef73f08764fbbb9cb9048a7R212

Was it done intentionally?

otemnov commented:
@nikita-bestuzhev-ct I've opened an issue https://github.com/actions/setup-dotnet/issues/739

Frulfump commented:
> @HarithaVattikuti, @priyagupta108, hello!
> 
> JFYI, the following `global.json`:
> ```
> {
>   "sdk": {
>     "version": "8.0",
>     "rollForward": "latestFeature"
>   }
> }
> ```
> now fails with:
> `Error: Cannot read properties of undefined (reading 'substring')`
> 
> See https://github.com/actions/setup-dotnet/pull/538/changes#diff-83766f6dc1065d750be380c25c707c33a1516e036ef73f08764fbbb9cb9048a7R212
> 
> Was it done intentionally?

8.0 isn't a valid version according to the docs for global.json https://learn.microsoft.com/en-us/dotnet/core/tools/global-json#version 


>version
>
>    Type: string
>
>The version of the .NET SDK to use.
>
>This field:
>
>    Requires the full version number, such as 10.0.100.
>    Doesn't support version numbers like 10, 10.0, or 10.0.x.
>    Doesn't have wildcard support.
>    Doesn't support version ranges.

cyberbossone-coder reviewed (changes requested):
This user cannot use or change this without permmission
